### PR TITLE
Add KL annealing to BNN VI training 

### PR DESCRIPTION
**Pull request by Moritesh**
# Add KL annealing to BNN VI training                                                                                                                                                                                                 
                                                                                                                                                                                                                                        
  > **Note:** This PR is based on the VI training enhancements from PR #132 (`num_particles`, gradient clipping, `kl_tau`, lr scheduler) and should be rebased & merged only after #132 lands on `develop`.                                       
                                                                                                                                                                                                                                        
  ## Summary                                                                                                                                                                                                                            
                                                                                                   
  - Introduces `kl_annealing_fraction` as a new VI `update_kwargs` parameter for `BayesianNeuralNetwork`. When set, the KL divergence term in the ELBO is linearly warmed up from `0 → 1` over the first `kl_annealing_fraction *       
  total_steps` training steps, then held at `1.0` for the remainder.
  - The annealing factor is threaded as a traced JAX scalar through the `model` and `guide` callables, so the warm-up schedule runs entirely inside `lax.scan` with no Python-loop overhead or XLA recompilation per step.              
  - Composes multiplicatively with the existing `kl_tau` temperature scale when both are active: `effective_scale = kl_base_factor * kl_annealing_factor`.                                                                              
  - `kl_annealing_fraction=None` and `=0.0` are both treated as inactive (no-op), validated at construction time.                                                                                                                       
                                                                                                                                                                                                                                        
  ## Test plan                                                                                                                                                                                                                          
                                                                                                                                                                                                                                        
  - [ ] Validation rejects values outside `[0, 1]`                                                 
  - [ ] `None` and `0.0` produce equivalent training behavior (both inactive paths)
  - [ ] Hypothesis-based test covers the full cross-product with existing VI options: `num_particles`, `gradient_clip_norm`, `kl_tau`, `lr_scheduler`  

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Optax optimizer support with optional LR scheduling, gradient clipping, and configurable num_particles.
  * Per-step KL annealing for variational inference with kl_tau and kl_annealing_fraction controls.

* **Chores**
  * Version bumped to 6.1.0.
  * Added Optax as a runtime dependency.

* **Tests**
  * Expanded test coverage for optimizer/scheduler combinations, VI training options, and KL-annealing behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: pybandits/model.py [hunk 1]
```diff
@@ -1532,30 +1599,36 @@ def create_update_model(self, batch_size: Optional[PositiveInt] = None) -> Calla
...
+        kl_scale = self._kl_scale_ctx
+
+        n_layers = len(self.model_params.bnn_layer_params)
         numerical_indices = self.feature_config.numerical_indices
         cat_configs = self.feature_config.categorical_features_configs
         has_embeddings = self.model_params.embedding_params is not None and len(cat_configs) > 0
 
-        def model(x: jax.Array, y: jax.Array):
-            N = x.shape[0]
+        def model(x: jax.Array, y: jax.Array, kl_annealing_factor=1.0):
```
Comment (shaharbar1) [inline pybandits/model.py]: Add Float01 typehint for kl_annealing_factor

File: pybandits/model.py [hunk 2]
```diff
@@ -1905,11 +2001,31 @@ def _run_svi_training_loop(self, x_jnp: jnp.ndarray, y_jnp: jnp.ndarray, n_sampl
             if not epoch_steps_list:
                 epoch_steps_list = [1]
 
+        total_steps = sum(epoch_steps_list)
+        kl_annealing_fraction = self._update_kwargs.get("kl_annealing_fraction")
+        kl_annealing_active = kl_annealing_fraction not in (None, 0.0)
+        kl_warmup_steps = max(1, int(np.ceil(kl_annealing_fraction * total_steps))) if kl_annealing_active else None
```
Comment (shaharbar1) [inline pybandits/model.py]: I think you can drop the np.ceil and just directly apply int().

File: pybandits/model.py [hunk 3]
```diff
@@ -1905,11 +2001,31 @@ def _run_svi_training_loop(self, x_jnp: jnp.ndarray, y_jnp: jnp.ndarray, n_sampl
...
+        raw_guide = method_config["guide"](_model, init_loc_fn=init_to_median)
+
+        # When kl_tau or KL annealing is active, wrap the guide with the same scale handler so
+        # both log p(z) and log q(z) are scaled equally → the KL term as a
+        # whole is multiplied by kl_base_factor * kl_annealing_factor.
+        kl_tau_active = self._update_kwargs.get("kl_tau") is not None
+        if kl_tau_active or kl_annealing_active:
+            kl_scale = self._kl_scale_ctx
+
+            def guide(x, y, kl_annealing_factor, *args, **kwargs):
```
Comment (shaharbar1) [inline pybandits/model.py]: Add typehints

File: pybandits/model.py [hunk 4]
```diff
@@ -1444,63 +1527,52 @@ def __setstate__(self, state: dict) -> None:
...
-            Backend-specific linear transform: ``(x, w, b) -> x @ w + b``.
-            For JAX: ``lambda x, w, b: jnp.dot(x, w) + b``.
-            For NumPy: ``lambda x, w, b: np.einsum("...i,...ij->...j", x, w) + b``.
-        backend : ModuleType
-            Array namespace — either ``jnp`` (JAX) or ``np`` (NumPy). Used for
-            ``backend.zeros`` and ``backend.concatenate`` in residual padding.
+        n_samples : PositiveInt
+            Number of data points in the current batch.
+        kl_annealing_factor : float or jax scalar, optional
+            Multiplicative factor in [0, 1]. Default 1.0 (no annealing).
```
Comment (shaharbar1) [inline pybandits/model.py]: I think you told me it is in (0, 1].

File: pybandits/model.py [hunk 5]
```diff
@@ -1918,59 +2034,70 @@ def _run_svi_training_loop(self, x_jnp: jnp.ndarray, y_jnp: jnp.ndarray, n_sampl
...
-        restore_best = self._update_kwargs.get("restore_best_svi_state", True)
         all_losses = []
-        best_loss = float("inf")
-        best_svi_state = svi_state
         pbar = trange(len(epoch_steps_list), desc="SVI", leave=False)
 
+        global_step_offset = 0
         for epoch_idx, epoch_steps in enumerate(epoch_steps_list):
-            svi_state, epoch_losses = _run_epoch(svi_state, epoch_steps)
+            if kl_annealing_active:
```
Comment (shaharbar1) [inline pybandits/model.py]: Instead of calculating steps, wouldn't it be better to precompute here the kl_factor and pass it to run_epoch?

File: tests/test_model.py [hunk 6]
```diff
@@ -1683,38 +1814,81 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
-    bnn = BayesianNeuralNetwork.cold_start(
-        n_features=n_features,
-        update_method=update_method,
-        update_kwargs=update_kwargs,
+# ---------------------------------------------------------------------------
+# KL annealing
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.parametrize("bad_value", [-0.1, 1.5, 2.0, -1.0])
```
Comment (shaharbar1) [inline tests/test_model.py]: Put all these tests on a designated class.

File: tests/test_model.py [hunk 7]
```diff
@@ -1683,38 +1814,81 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+def test_kl_annealing_fraction_range_validation(bad_value: float):
+    """Out-of-range kl_annealing_fraction raises ValueError at construction time."""
+    with pytest.raises(ValueError, match="kl_annealing_fraction"):
```
Comment (shaharbar1) [inline tests/test_model.py]: Please avoid hard-coded constants in the tests to make them easily-maintainable.
Try using pytest.mark.parameterize and hypothesis given for better coverage.

**Review by shaharbar1 — CHANGES_REQUESTED** [review]

File: pybandits/model.py (hunk 1 above)
Comment (Moritesh) [reply pybandits/model.py]: I've annotated as Union[Float01, jax.Array] = 1.0 rather than bare Float01 because this argument is consumed inside lax.scan where it's a traced jax.Array, not a Python float. Same shape as the _kl_scale_ctx(n_samples) annotation fix in your other comment.

File: pybandits/model.py (hunk 2 above)
Comment (Moritesh) [reply pybandits/model.py]: I'd actually prefer to keep np.ceil here. With plain int(...) truncation we'd silently give fewer warmup steps than the user requested (e.g. 0.05 * 195 = 9.75 rounds to 9 steps instead of 10), and for small fractions on small training runs it can truncate to 0, leaning entirely on the max(1, ...) floor. Rounding up feels like the more honest reading of "warm up over kl_annealing_fraction * total_steps steps". Can drop it if you feel strongly though.

File: pybandits/model.py (hunk 4 above)
Comment (Moritesh) [reply pybandits/model.py]: Indeed, fixing in documentation

File: pybandits/model.py (hunk 5 above)
Comment (Moritesh) [reply pybandits/model.py]: Good thinking, it's indeed a cleaner design

File: tests/test_model.py (hunk 6 above)
Comment (Moritesh) [reply tests/test_model.py]: Done. Although worth noting that this is currently the only Test* class in tests/test_model.py - the rest of the file is module-level functions with # --- banner comments (see the categorical-embeddings, MCMC, and VI sections). I noticed that other test files in the repo do use Test* classes though.

File: pybandits/model.py (hunk 2 above)
Comment (shaharbar1) [reply pybandits/model.py]: I agree with you

File: tests/test_model.py [hunk 8]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+            BayesianNeuralNetwork.cold_start(
+                n_features=self.N_FEATURES,
+                update_kwargs={"kl_annealing_fraction": bad_value},
+            )
+
+    @pytest.mark.parametrize("kl_annealing_fraction", [None, 0.0])
+    def test_kl_annealing_inactive_no_scale_on_prior_sites(self, kl_annealing_fraction):
+        """Inactive values (None, 0.0) gate _kl_scale_ctx to nullcontext, so prior
+        sample sites carry no scale annotation in the model trace."""
+        import jax.numpy as jnp
```
Comment (shaharbar1) [inline tests/test_model.py:1843]: avoid local imports

File: pybandits/model.py
Comment (shaharbar1) [inline pybandits/model.py:1]: Cool and neat job @Moritesh

File: tests/test_model.py [hunk 9]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
-    context = np.random.rand(n_samples, n_features).astype(np.float32)
-    rewards = _make_random_rewards(n_samples)
+# ---------------------------------------------------------------------------
+# KL annealing
+# ---------------------------------------------------------------------------
 
-    original_forward_layers = BaseBayesianNeuralNetwork._forward_layers
-    call_count = 0
+# Boundary-derived invalid fractions covering both sides of the (0, 1] range.
+_INVALID_KL_ANNEALING_FRACTIONS = (-1e-6, -0.1, -1.0, 1.0 + 1e-6, 1.5, 2.0)
```
Comment (shaharbar1) [inline tests/test_model.py:1822]: Why is this outside the class and n_features is inside? you can put both inside.

File: tests/test_model.py [hunk 10]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
-    assert call_count == ref, "sample_proba must call _forward_layers"
+class TestKLAnnealing:
+    """Tests for the kl_annealing_fraction VI knob on BayesianNeuralNetwork."""
 
-    # Check update calls _forward_layers
-    call_count = 0
-    with patch.object(BaseBayesianNeuralNetwork, "_forward_layers", tracking_forward_layers):
-        bnn.update(context=context, rewards=rewards)
-    assert call_count >= ref, "update must call _forward_layers"
+    N_FEATURES = 2
```
Comment (shaharbar1) [inline tests/test_model.py:1828]: I don't think you need capital letters when it's a class var

File: tests/test_model.py [hunk 11]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+                update_kwargs={"kl_annealing_fraction": bad_value},
+            )
+
+    @pytest.mark.parametrize("kl_annealing_fraction", [None, 0.0])
+    def test_kl_annealing_inactive_no_scale_on_prior_sites(self, kl_annealing_fraction):
+        """Inactive values (None, 0.0) gate _kl_scale_ctx to nullcontext, so prior
+        sample sites carry no scale annotation in the model trace."""
+        import jax.numpy as jnp
+
+        n_samples = 10
```
Comment (shaharbar1) [inline tests/test_model.py:1845]: make n_samples a test argument

File: tests/test_model.py [hunk 12]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+        context = jnp.asarray(np.random.rand(n_samples, self.N_FEATURES).astype(np.float32))
```
Comment (shaharbar1) [inline tests/test_model.py:1846]: use make_binary_rewards and make_context_matrix from tests/utils.py

File: tests/test_model.py (hunk 11 above)
Comment (shaharbar1) [reply tests/test_model.py:1845]: if possible, use @given for testing multiple values

File: tests/test_model.py [hunk 13]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+                f"factor={factor}: site {name!r} has scale={site.get('scale')!r}, expected {factor}."
+            )
+
+    def test_kl_annealing_interacts_with_kl_tau(self):
+        """kl_tau alone vs kl_tau + annealing produce distinct loss trajectories."""
+        random_seed = int(np.random.randint(0, 2**31 - 1))
+        kl_tau = float(np.random.uniform(0.01, 1.0))
+        # Lower bound 0.3: at num_steps=10, smaller fractions ramp the warmup
+        # window in too few steps to produce a divergent trajectory.
+        kl_annealing_fraction = float(np.random.uniform(0.3, 1.0))
```
Comment (shaharbar1) [inline tests/test_model.py:1899]: use hypothesis.given for random_seed , kl_tau , kl_annealing_fraction . 
BTW, I don't think random_seed is a must here, you can avoid passing it.

File: tests/test_model.py [hunk 14]
```diff
@@ -1683,38 +1814,129 @@ def test_bnn_vi_update_with_categorical_features_updates_embeddings(
...
+
+        # Both should complete without error and produce valid histories
+        assert bnn_tau_only._approx_history is not None
+        assert bnn_tau_anneal._approx_history is not None
+        # The KL weighting differs during warmup, so losses should diverge
+        assert not np.allclose(bnn_tau_only._approx_history, bnn_tau_anneal._approx_history)
+
+    def test_kl_annealing_fraction_serialization_round_trip(self):
+        """kl_annealing_fraction survives a model_dump / reconstruction round-trip."""
+        kl_annealing_fraction = float(np.random.uniform(0.01, 1.0))
```
Comment (shaharbar1) [inline tests/test_model.py:1931]: @given

**Review by shaharbar1 — CHANGES_REQUESTED** [review]
